server/src/main/java/org/elasticsearch/cluster/routing/allocation/allocator/DesiredBalanceReconciler.java

```diff
@@ -64,12 +66,29 @@ public class DesiredBalanceReconciler {
         Setting.Property.NodeScope
     );
 
+    private final FrequencyCappedAction logReconciliationMetrics;
     private final FrequencyCappedAction undesiredAllocationLogInterval;
     private double undesiredAllocationsLogThreshold;
     private final NodeAllocationOrdering allocationOrdering = new NodeAllocationOrdering();
     private final NodeAllocationOrdering moveOrdering = new NodeAllocationOrdering();
 
+    // stats
+    /**
+     * Number of unassigned shards during last reconciliation
+     */
+    protected final AtomicInteger unassignedShards = new AtomicInteger();
+    /**
+     * Total number of assigned shards during last reconciliation
+     */
+    protected final AtomicInteger totalAllocations = new AtomicInteger();
+    /**
+     * Number of assigned shards during last reconciliation that are not allocated on desired node and need to be moved
+     */
+    protected final AtomicInteger undesiredAllocations = new AtomicInteger();
+
     public DesiredBalanceReconciler(ClusterSettings clusterSettings, ThreadPool threadPool) {
+        this.logReconciliationMetrics = new FrequencyCappedAction(threadPool);
+        this.logReconciliationMetrics.setMinInterval(TimeValue.timeValueMinutes(30));
```

Comment:
nit Would it make sense to re-use UNDESIRED_ALLOCATIONS_LOG_INTERVAL_SETTING or define a new setting for this?

Reply:
This one is intended for internal consumption, the other one is to warn users. So I would prefer not to share implementation details.
I am also hoping to replace log based collection with APM as soon as it is available